controller/WFCORE-4770-boot-cli-script.adoc

```diff
@@ -0,0 +1,283 @@
+
+> If the EAP container launch script detects that some configuration is required, it starts the server in admin-mode.
+Once configuration is over, the server is reloaded  for the rare cases) into normal mode. All
+configurations aspects are handled internally by the EAP container. The extensions hook allows users to add their own
+custom CLI operations to the global server configuration.”
+
+It means that we now have a `CONFIG` phase (where the server runs in admin-only mode during the
+boot/configuration), followed by a `RUNNING` phase.
+
+New env vars:
+* `EXECUTE_BOOT_SCRIPT_INVOKER` - to control if the CLI boot time hook is to be used. Set to `true` in the builder image.
+If `false`, we fall back on the existing way of configuring the server where we launch embedded server session(s) in CLI
+to configure the server before starting the server in normal mode.
+* `EXECUTE_BOOT_SCRIPT_INVOKER_TIMEOUT` - to control the timeout to wait for the server to generate the marker file
+(if needed, @see sequence with extensions).
+
+Detailed sequences:
+
+. Standard use-case, no user extensions installed:
+* Preconfigure and configure are called on all launch modules (same as today).
+* The server is started in admin-only mode with the generated script passed as a system property, as well as the
+marker directory specified.
+* Any `WARN` messages during CLI execution are printed
+* If error, the error message is displayed and the server aborts
+* The server reloads itself in normal mode
+. Optional use-case, user extensions installed:
+* The server is started in admin mode with following system properties set: CLI script, no reload and marker directory.
+** *NB: The server is started in background (that is already the case today).*
+* Once CLI execution done, the server writes the marker file.
+** *NB: The server IS NOT yet reloaded in normal mode.*
+* The shell script waits for the marker file to appear. If it doesn’t appear after a configurable timeout
+(default 30secs), the script aborts.
+* If the marker appeared and contains `success`, the script configures the extensions (pure bash and remote CLI).
+If this fails, the script is aborted.
+* If an error occurred during the above step, the server is shutdown.
+* Otherwise,a CLI process is started, connects to remote running server, executes the extensions generated script (if any).
+** If an error occurred during this CLI execution, the server is shutdown.
+** If success, the following logic applies:
+*** If restart is required. `:shutdown(restart=true)` is executed which causes the launch script to restart the server
+with the original arguments. Following the restart of the admin-only server it is reloaded into normal mode.
+*** Otherwise the server is reloaded into normal mode.
+* The CLI process terminates.
+
+==== Impact on readiness probe and Operator
+
+While running in admin-mode the server shouldn’t be advertised as ready.
```

Comment:
This feature has an impact on our existing [EAP probe](https://github.com/jboss-container-images/jboss-eap-modules/blob/66f3a7d08876a8e90a337bde5fb2abf3b29bf78c/os-eap-probes/common/added/probes/probe/eap/dmr.py#L38).
The probe UP if the `server-state` is `running` but does not look at the value of the `running-mode` attribute (which can be either `NORMAL` or `ADMIN_ONLY`). 

The probe would have to be updated when this feature is activated.

Reply:
@jmesnil I added ^^ to the proposal